src/main/java/io/cdap/delta/bigquery/BigQueryUtils.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.delta.bigquery;
+
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.EncryptionConfiguration;
+import com.google.cloud.bigquery.FieldValue;
+import com.google.cloud.bigquery.FieldValueList;
+import com.google.cloud.bigquery.Job;
+import com.google.cloud.bigquery.JobId;
+import com.google.cloud.bigquery.JobInfo;
+import com.google.cloud.bigquery.QueryJobConfiguration;
+import com.google.cloud.bigquery.TableId;
+import com.google.cloud.bigquery.TableResult;
+import io.cdap.delta.api.DeltaTargetContext;
+import io.cdap.delta.api.SourceTable;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+
+public final class BigQueryUtils {
+  private static final int MAX_LENGTH = 1024;
+  // according to big query dataset and table naming convention, valid name should only contain letters (upper or
+  // lower case), numbers, and underscores
+  private static final String VALID_NAME_REGEX = "[\\w]+";
+  private static final String INVALID_NAME_REGEX = "[^\\w]+";
+
+  private BigQueryUtils() {
+  }
+
+  static long getMaximumExistingSequenceNumber(DeltaTargetContext context, String project, @Nullable String datasetName,
+                                               BigQuery bigQuery, EncryptionConfiguration encryptionConfiguration) {
+    StringBuilder builder = new StringBuilder();
+    builder.append("SELECT MAX(max_sequence_num) FROM (");
+    Set<SourceTable> allTables = context.getAllTables();
+    List<String> maxSequenceNumQueryPerTable = new ArrayList<>();
+    for (SourceTable table : allTables) {
+      TableId tableId = TableId.of(project,
+                                   datasetName != null ? normalize(datasetName) : normalize(table.getDatabase()),
+                                   normalize(table.getTable()));
+      if (bigQuery.getTable(tableId) != null) {
+        maxSequenceNumQueryPerTable.add(String.format("SELECT MAX(_sequence_num) as max_sequence_num FROM %s.%s.%s",
+                                                      tableId.getProject(), tableId.getDataset(), tableId.getTable()));
```

Comment:
do we need the project ? getMaximumSequenceNumberForTable does't use project, shall we make them consistent?

Reply:
Its not required. Removed.